signatures/pdf_sign_external_google_cloud_kms.go

```diff
@@ -0,0 +1,566 @@
+	copy(pdfData[byteRange[1]:byteRange[2]], []byte(sig))
+
+	// Write output file.
+	if err := ioutil.WriteFile(outputPath, pdfData, os.ModePerm); err != nil {
+		log.Fatalf("Fail: %v\n", err)
+	}
+
+	log.Printf("PDF file successfully signed. Output path: %s\n", outputPath)
+}
+
+// generateSignedFile generates a signed version of the input PDF file using the
+// specified signature handler.
+func generateSignedFile(inputPath string, handler model.SignatureHandler) ([]byte, *model.PdfSignature, error) {
+	// Create reader.
+	file, err := os.Open(inputPath)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer file.Close()
+
+	reader, err := model.NewPdfReader(file)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create appender.
+	appender, err := model.NewPdfAppender(reader)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create signature.
+	signature := model.NewPdfSignature(handler)
+	signature.SetName("GOOGLECLOUD KMS Sign")
+	signature.SetReason("TestGCKMSSignature")
+	signature.SetDate(now, "")
+
+	if err := signature.Initialize(); err != nil {
+		return nil, nil, err
+	}
+
+	// Create signature field and appearance.
+	opts := annotator.NewSignatureFieldOpts()
+	opts.FontSize = 10
+	opts.Rect = []float64{10, 25, 75, 60}
+
+	field, err := annotator.NewSignatureField(
+		signature,
+		[]*annotator.SignatureLine{
+			annotator.NewSignatureLine("Name", "John Doe"),
+			annotator.NewSignatureLine("Date", "2023.01.01"),
+			annotator.NewSignatureLine("Reason", "GC KMS Sing Test"),
+		},
+		opts,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	field.T = core.MakeString("External signature")
+
+	if err = appender.Sign(1, field); err != nil {
+		return nil, nil, err
+	}
+
+	// Write PDF file to buffer.
+	pdfBuf := bytes.NewBuffer(nil)
+	if err = appender.Write(pdfBuf); err != nil {
+		return nil, nil, err
+	}
+
+	return pdfBuf.Bytes(), signature, nil
+}
+
+// getExternalSignatureAndSign simulates an external service which signs the specified
+// PDF file and returns its pdf signed bytes and signature.
+func getExternalSignatureAndSign(inputPath, credPath, keyName string) ([]byte, []byte, error) {
+	gcKmsSign, err := GcKmsExternalSigner(credPath, keyName)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer gcKmsSign.client.Close()
+
+	now := time.Now()
+
+	// Initialize X509 certificate template.
+	template := x509.Certificate{
+		SerialNumber: new(big.Int),
+		Subject: pkix.Name{
+			CommonName:   "UniDOC",
+			Organization: []string{"Test Company"},
+		},
+		NotBefore:          now.Add(-time.Hour),
+		NotAfter:           now.Add(time.Hour * 24 * 365),
+		PublicKeyAlgorithm: x509.RSA,
+		SignatureAlgorithm: x509.SHA256WithRSA,
+		KeyUsage:           x509.KeyUsageDigitalSignature,
+	}
+
+	// We sign certificate using external signer that implement `crypto.Signer`.
+	certData, err := x509.CreateCertificate(rand.Reader, &template, &template, gcKmsSign.getPublicKey(), gcKmsSign.signer)
+	if err != nil {
+		panic(err)
```

Comment:
should return/pass the error up, also below

Reply:
Updated at 9df3ed013f0019217c0977de3a173ea41ea1289a